### DISCUSSION TITLE
Can we use spotlightClicks without a controlled/stepIndex mode?

### DISCUSSION THREAD

**Discussion by jimmythecode** [gilbarbara/react-joyride#665] [Q&A] [answered]
I've set up a simple demo in [sandbox](https://codesandbox.io/s/joyride-basic-question-how-to-spotlight-click-to-next-step-gtytr?file=/src/Joyride.js:575-590) with a basic set-up. Step 4 uses Button1 as a target, and I've set the prop for Step 4 ```spotlightClicks: true```. This works, and allows the button to be clicked, but then it doesn't move to the next step. 

Do I need to switch to controlled mode to use this feature or can it be achieved with the basic set up?

Also (bonus question), if using the spotlightClicks feature to click a button to another page (using react router), can Joyride move from my id="#button1" element in Step 4, to a new element (eg id="button2"), on another page? I'm wondering if there's a point between closing one route and opening another where Joyride has no target and this might break the component.

**Comment by gilbarbara** [comment, marked as answer]
hey @jimmythecode 

The `spotlightClicks` works with both modes, uncontrolled or controlled.
However, this library only has control of the elements it creates, so clicking a button in your code won't have any direct effect on the tour...

If you need the tour to work in different contexts (routes in your case) then you'll need to use the controlled mode.

You'll need a handler in this button to stop the tour, increase the stepIndex, and after the route loads (and your next target is visible) start the tour again.

**Reply by jimmythecode** [reply]
Excellent. Thanks for the advice.

Given that, as you say "However, this library only has control of the elements it creates", I'm struggling to understand how the ```spotlightClicks``` could work in uncontrolled mode? Do you mean there's a way to create a Joyride element to place within the spotlight? Otherwise I can't see how you could use a spotlightClick and move to the next step?

**Reply by gilbarbara** [reply]
The element that is exposed through the spotlight could have tons of different functions besides route navigation.
It could show another UI element on the same page or perform a background call, whatever.

This is a general-purpose library, route navigation is just one of many needs

**Reply by jimmythecode** [reply]
Still though, in any situation I am trying to imagine in uncontrolled mode, in any situation the user:

1. Clicks through the spotlight
2. An event happens, and (eg) something changes on the page.
3. The tour stays static.

The user would then have to click 'Next' to progress the tour. But it would make sense in my tour. I think typically the user would expect to click through the spotlight, and the step would automatically change to the next one in the array.

One thing that I'm struggling with, is that when I click through the spotlight, nothing is picked up in the Joyride callback, so I can't trigger a change from Joyride. For example, I'd like to navigate to a new page, and then change the steps state so that I have a new array of steps for this new page. I think I need to restart Joyride for this to work though. 

**EDIT: I thought I could achieve it with:** ```if (... type === ACTIONS.UPDATE)``` **but when I try to setRunningState to false it has no effect?**

**Reply by gilbarbara** [reply]
I could have a big area in the spotlight that can be interacted within the visible part that doesn't need to change the tour's step. As I said, this library doesn't change or interact with your own components, **it is just a layer on top of your content.**
There are thousands of different users with different needs and that's why there's the controlled mode...

There's no intrusive (and hacky) way this library could know what's happening with your code. This will never change.
If you need that an action within your app to change Joyride's state, you'll need to update it by yourself.

**Reply by jimmythecode** [reply]
Yeah that makes sense to me. I've a question that touches on what we've been discussing. If we use ```spotlightClicks: true``` and I click a button within, is there a way to progress a step forward purely within Joyride (eg using its callback function)?

I've created a codesandbox to demonstrate:
https://codesandbox.io/s/joyride-controlled-demos-hwojx?file=/src/Tutorials/SpotlightInternal.js

If you go to the "spotlightClicks Internal Handling" option in the menu it will explain better. Thanks!

**Reply by gilbarbara** [reply]
No

**Reply by gilbarbara** [reply]
This `spotlightClicks` prop only makes the spotlight area not capture the clicks using the CSS `pointer-events: none` statement. That's the only change.

**It's impossible for this library to watch/interact/respond to interactions in your app. The only thing it knows is the target class/element. Period.**

**Reply by jimmythecode** [reply]
Nooo, sorry, I mustn't have been clear. I wasn't asking it to interact with my app. I want it to allow me to click an element (as it does with ```spotlightClicks:true```, so my app does its interactions itself. It's just I want to know if Joyride can update **itself** - for example, it updates its own state when I click within the spotlight. I could then use the callback to update my own state.

So when I click away from a tooltip, Joyride responds to this. I wondered if it can do the same when I click within the tooltip.

**Reply by gilbarbara** [reply]
Since the spotlight doesn't capture clicks because of pointer-events property it won't respond to any interactions.
